### PR TITLE
MAINT: Add and check type hints

### LINKED ISSUE #994: ENH: Type checking
@hoechenberger I saw in a recent PR that @drammock added some missing type hints. I am not very experienced with type checking but this repo seems pretty well documented -- worth adding `mypy` to pre-commit or at least a CI run? I briefly tried both `mypy` and `pyright` and `mypy` was a little bit faster and used already by MNE so I'm inclined to try that.

I already see some annoyingness with TypedDict not subclassing dict but I'll look into that if type checking does seem worthwhile.

### PR DESCRIPTION
**Pull request by larsoner**
Initial push to see if I broke anything. 

- [x] Changelog has been updated (`docs/source/changes.md`)
- [x] I still had a function with no type annotations so need to tell mypy to be picky about those, too.

Closes #994

### CODE REVIEW

**Comment by larsoner** [comment]
Green @hoechenberger !

File: docs/source/examples/gen_examples.py [hunk 1]
```diff
@@ -228,8 +231,8 @@ def _gen_demonstrated_funcs(example_config_path: Path) -> dict:
     source_str = f"## Dataset source\n\nThis dataset was acquired from [{url}]({url})\n"
 
     if "openneuro" in options:
-        for key in ("include", "exclude"):
-            options[key] = options.get(key, [])
+        options["include"] = options.get("include", [])
+        options["exclude"] = options.get("exclude", [])
```
Comment (drammock) [inline docs/source/examples/gen_examples.py:235]: idle curiosity: Is there any difference / advantage between this idiom and 

```suggestion
        _ = options.setdefault("include", [])
        _ = options.setdefault("exclude", [])
```

(other than the fact that `setdefault` returns the value)?

Comment (larsoner) [reply docs/source/examples/gen_examples.py:235]: Probably not I just never really use `setdefault` :)

**Comment by larsoner** [comment]
@hoechenberger let me know if you want to look otherwise I'll go ahead and merge!

**Comment by hoechenberger** [comment]
Hey, I won't have time to review until Friday or Saturday 

will that suffice?

**Comment by larsoner** [comment]
Sure!

**Comment by larsoner** [comment]
@hoechenberger okay for me to fix the merge conflict and merge?

**Review by hoechenberger — APPROVED** [review]

**Comment by hoechenberger** [comment]
Excellent work, @larsoner

**Pull request merged**